src/parser/Expression.ts

```diff
@@ -645,5 +645,112 @@ export class CallfuncExpression implements Expression {
+        let [consequentExpressions, consequentVarExpressions, consequentUniqueNames] = this.getExpressionInfo(this.consequent);
+        let [alternateExpressions, alternateVarExpressions, alternateUniqueNames] = this.getExpressionInfo(this.alternate);
+
+        let allExpressions = [...testExpressions, ...consequentExpressions, ...alternateExpressions];
+        let allUniqueVarNames = [...new Set([...testUniqueNames, ...consequentUniqueNames, ...alternateUniqueNames])];
+
+        let mutatingExpressions = allExpressions.filter(e => e instanceof CallExpression || e instanceof CallfuncExpression || e instanceof DottedGetExpression);
+
+        // - TODO - it doesn't look like we need to manipulate the variable names
+        // we can assign m on scope, and it should be fine
+
+        //FIXME - need to ensure I'm using SourceNode's correctly here - not sure how to do that yet
+        if (mutatingExpressions.length > 0) {
+            //we need to do a scope-safe ternary operation
+            let scope = '{';
+            // eslint-disable-next-line no-return-assign
+            allUniqueVarNames.forEach(name => scope += `\n"${name}": name`);
+            scope += '}';
+
+            result.push(`bslib_scopeSafeTernary(${this.test.transpile(state)}, ${scope},`);
+            result = result.concat(this.getScopedFunction(state, this.consequent, consequentUniqueNames));
+            result = result.concat(this.getScopedFunction(state, this.alternate, alternateUniqueNames));
+            result.push(')');
+        } else {
+            const ternaryCall = `bslib_simpleTernary(${this.test.transpile(state)}, ${this.consequent.transpile(state)}, ${this.alternate.transpile((state))})`;
+
+            result.push(new SourceNode(this.range.start.line + 1, this.range.start.character, state.pathAbsolute, ternaryCall));
+        }
+        return result;
+    }
+
+    private getExpressionInfo(expression: Expression): [Expression[], Expression[], string[]] {
+        const expressions = getAllExpressions(expression);
+        const varExpressions = expressions.filter(e => e instanceof VariableExpression) as VariableExpression[];
+        const uniqueVarNames = [...new Set(varExpressions.map(e => e.name.text))];
+        return [expressions, varExpressions, uniqueVarNames];
+    }
+
+    private getScopedFunction(state: TranspileState, alternate: Expression, scopeVarNames: string[]): Array<string| SourceNode> {
+        let result = [];
+        let text = 'function(scope)\n';
+        scopeVarNames.forEach(name => {
+            text += `${name} = scope.${name}\n`;
+        });
+        result.push(text);
+        result = result.concat(alternate.transpile(state));
+        result.push('end function\n');
+        return result;
+    }
+}
+
+/**
+ * Walks an expression, getting all nested expressions
+ * @param expression
+ * @param expressions
+ */
+function getAllExpressions(expression: Expression, expressions: Expression[] = []): Expression[] {
```

Comment:
Is there a way we can handle this without needing to hardcode every known expression type? Like perhaps we make each expression EXTEND Expression instead of implement the expression interface, then we can test for Expression in the prototype chain. I'm just worried about adding new expression types over time and not remembering to update this function to account for that.

Reply:
yes - I put a comment in here saying I think each expression should implement this method. Glad you concur. I'll make it so.